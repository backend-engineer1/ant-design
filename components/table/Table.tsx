import * as React from 'react';
import { Summary, EXPAND_COLUMN } from 'rc-table';
import Column from './Column';
import ColumnGroup from './ColumnGroup';
import InternalTable from './InternalTable';
import type { TableProps } from './InternalTable';
import type { RefTable } from './interface';
import {
  SELECTION_ALL,
  SELECTION_COLUMN,
  SELECTION_INVERT,
  SELECTION_NONE,
} from './hooks/useSelection';

function Table<RecordType extends object = any>(
  props: TableProps<RecordType>,
  ref: React.Ref<HTMLDivElement>,
) {
  const renderTimesRef = React.useRef(0);
  renderTimesRef.current += 1;

  return <InternalTable<RecordType> {...props} ref={ref} _renderTimes={renderTimesRef.current} />;
}

const ForwardTable = React.forwardRef(Table) as any as RefTable & {
  SELECTION_COLUMN: typeof SELECTION_COLUMN;
  EXPAND_COLUMN: typeof EXPAND_COLUMN;
  SELECTION_ALL: 'SELECT_ALL';
  SELECTION_INVERT: 'SELECT_INVERT';
  SELECTION_NONE: 'SELECT_NONE';
  Column: typeof Column;
  ColumnGroup: typeof ColumnGroup;
  Summary: typeof Summary;
};

ForwardTable.SELECTION_COLUMN = SELECTION_COLUMN;
ForwardTable.EXPAND_COLUMN = EXPAND_COLUMN;
ForwardTable.SELECTION_ALL = SELECTION_ALL;
ForwardTable.SELECTION_INVERT = SELECTION_INVERT;
ForwardTable.SELECTION_NONE = SELECTION_NONE;
ForwardTable.Column = Column;
ForwardTable.ColumnGroup = ColumnGroup;
ForwardTable.Summary = Summary;

export default ForwardTable;
