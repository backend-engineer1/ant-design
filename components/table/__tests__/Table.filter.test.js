import React from 'react';
import { render, mount } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';
import Table from '..';

describe('Table.filter', () => {
  const filterFn = (value, record) => record.name.indexOf(value) !== -1;
  const column = {
    title: 'Name',
    dataIndex: 'name',
    filters: [
      { text: 'Boy', value: 'boy' },
      { text: 'Girl', value: 'girl' },
      {
        text: 'Title',
        value: 'title',
        children: [
          { text: 'Designer', value: 'designer' },
          { text: 'Coder', value: 'coder' },
        ],
      },
    ],
    onFilter: filterFn,
  };

  const data = [
    { key: 0, name: 'Jack' },
    { key: 1, name: 'Lucy' },
    { key: 2, name: 'Tom' },
    { key: 3, name: 'Jerry' },
  ];

  function createTable(props) {
    return (
      <Table
        columns={[column]}
        dataSource={data}
        pagination={false}
        {...props}
      />
    );
  }

  function renderedNames(wrapper) {
    return wrapper.find('TableRow').map(row => row.props().record.name);
  }

  it('renders filter correctly', () => {
    const wrapper = render(createTable());

    expect(renderToJson(wrapper)).toMatchSnapshot();
  });

  it('renders menu correctly', () => {
    const wrapper = mount(createTable());
    const dropdownWrapper = render(wrapper.find('Trigger').node.getComponent());
    expect(renderToJson(dropdownWrapper)).toMatchSnapshot();
  });

  it('renders radio filter correctly', () => {
    const wrapper = mount(createTable({
      columns: [{
        ...column,
        filterMultiple: false,
      }],
    }));
    const dropdownWrapper = render(wrapper.find('Trigger').node.getComponent());
    expect(renderToJson(dropdownWrapper)).toMatchSnapshot();
  });

  it('renders custom content correctly', () => {
    const filter = (
      <div className="custom-filter-dropdown">
        custom filter
      </div>
    );
    const wrapper = mount(createTable({
      columns: [{
        ...column,
        filterDropdown: filter,
      }],
    }));

    const dropdownWrapper = render(wrapper.find('Trigger').node.getComponent());
    expect(renderToJson(dropdownWrapper)).toMatchSnapshot();
  });

  it('can be controlled by filterDropdownVisible', () => {
    const wrapper = mount(createTable({
      columns: [{
        ...column,
        filterDropdownVisible: true,
      }],
    }));
    const dropdown = wrapper.find('Dropdown').first();

    expect(dropdown.props().visible).toBe(true);
    wrapper.setProps({ columns: [{
      ...column,
      filterDropdownVisible: false,
    }] });
    expect(dropdown.props().visible).toBe(false);
  });

  it('fires change event when visible change', () => {
    const handleChange = jest.fn();
    const wrapper = mount(createTable({
      columns: [{
        ...column,
        onFilterDropdownVisibleChange: handleChange,
      }],
    }));

    wrapper.find('Dropdown').first().simulate('click');

    expect(handleChange).toBeCalledWith(true);
  });

  it('can be controlled by filteredValue', () => {
    const wrapper = mount(createTable({
      columns: [{
        ...column,
        filteredValue: ['Lucy'],
      }],
    }));

    expect(wrapper.find('tbody tr').length).toBe(1);
    wrapper.setProps({ columns: [{
      ...column,
      filteredValue: [],
    }] });
    expect(wrapper.find('tbody tr').length).toBe(4);
  });

  it('can be controlled by filteredValue null', () => {
    const wrapper = mount(createTable({
      columns: [{
        ...column,
        filteredValue: ['Lucy'],
      }],
    }));

    expect(wrapper.find('tbody tr').length).toBe(1);
    wrapper.setProps({ columns: [{
      ...column,
      filteredValue: null,
    }] });
    expect(wrapper.find('tbody tr').length).toBe(4);
  });

  it('fires change event', () => {
    const handleChange = jest.fn();
    const wrapper = mount(createTable({ onChange: handleChange }));
    const dropdownWrapper = mount(wrapper.find('Trigger').node.getComponent());

    dropdownWrapper.find('MenuItem').first().simulate('click');
    dropdownWrapper.find('.confirm').simulate('click');

    expect(handleChange).toBeCalledWith({}, { name: ['boy'] }, {});
  });

  it('three levels menu', () => {
    const filters = [
      { text: 'Upper', value: 'Upper' },
      { text: 'Lower', value: 'Lower' },
      {
        text: 'Level2',
        value: 'Level2',
        children: [
          { text: 'Large', value: 'Large' },
          { text: 'Small', value: 'Small' },
          {
            text: 'Level3',
            value: 'Level3',
            children: [
              { text: 'Black', value: 'Black' },
              { text: 'White', value: 'White' },
              { text: 'Jack', value: 'Jack' },
            ],
          },
        ],
      },
    ];
    const wrapper = mount(createTable({
      columns: [{
        ...column,
        filters,
      }],
    }));
    const dropdownWrapper = mount(wrapper.find('Trigger').node.getComponent());
    dropdownWrapper.find('.ant-dropdown-menu-submenu-title').at(0).simulate('mouseEnter');
    dropdownWrapper.find('.ant-dropdown-menu-submenu-title').at(1).simulate('mouseEnter');
    dropdownWrapper.find('MenuItem').last().simulate('click');
    dropdownWrapper.find('.confirm').simulate('click');
    expect(renderedNames(wrapper)).toEqual(['Jack']);
  });
});
