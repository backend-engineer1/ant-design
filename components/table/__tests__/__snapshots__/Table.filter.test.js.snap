// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.filter override custom filter correctly 1`] = `
<div
  className="ant-dropdown-custom-view"
  id="customFilter"
>
  <span
    id="setSelectedKeys"
    onClick={[Function]}
  >
    setSelectedKeys
  </span>
  <span
    id="confirm"
    onClick={[Function]}
  >
    Confirm
  </span>
  <span
    id="reset"
    onClick={[Function]}
  >
    Reset
  </span>
</div>
`;

exports[`Table.filter renders custom content correctly 1`] = `
<div>
  <div
    class="ant-dropdown  ant-dropdown-placement-bottomLeft  ant-dropdown-hidden"
  >
    <div>
      <div
        class="custom-filter-dropdown"
      >
        custom filter
      </div>
    </div>
  </div>
</div>
`;

exports[`Table.filter renders custom filter icon correctly 1`] = `
<span
  class="ant-table-filter-icon ant-dropdown-trigger"
  style=""
  title="Filter menu"
>
  filtered
</span>
`;

exports[`Table.filter renders custom filter icon correctly 2`] = `
<span
  class="ant-table-filter-icon ant-dropdown-trigger"
  style=""
  title="Filter menu"
>
  unfiltered
</span>
`;

exports[`Table.filter renders filter correctly 1`] = `
<div
  class="ant-table-wrapper"
>
  <div
    class="ant-spin-nested-loading"
  >
    <div
      class="ant-spin-container"
    >
      <div
        class="ant-table ant-table-default ant-table-scroll-position-left"
      >
        <div
          class="ant-table-content"
        >
          <div
            class="ant-table-body"
          >
            <table
              class=""
            >
              <colgroup>
                <col />
              </colgroup>
              <thead
                class="ant-table-thead"
              >
                <tr>
                  <th
                    class="ant-table-column-has-filters"
                  >
                    <span>
                      Name
                      <i
                        class="anticon anticon-filter ant-dropdown-trigger"
                        title="Filter menu"
                      >
                        <svg
                          aria-hidden="true"
                          class=""
                          data-icon="filter"
                          fill="currentColor"
                          height="1em"
                          viewBox="0 0 1024 1024"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M890.1 176.6c-12.8-31-42.8-51-76.4-51H212c-33.5 0-63.5 20-76.4 51-12.9 31-5.9 66.3 17.8 90.1l273 263.3v262c0 9.3 3.6 18.3 10.1 25l125.2 129.5c6.9 7.2 16.3 11 25.9 11 4.5 0 9.1-.9 13.5-2.6 13.6-5.5 22.5-18.7 22.5-33.4V528.3l248.5-261.5c23.8-23.8 30.8-59.1 18-90.2zm-68.9 39.2l-259 272.1c-6.8 6.8-10.5 15.9-10.5 25.5v322l-53.2-58V515c0-9.5-3.8-18.7-10.5-25.4L204.4 215.8c-4.2-4.2-3.5-8.6-2.3-11.6 1.2-3 4-6.6 9.9-6.6h601.7c5.9 0 8.6 3.6 9.9 6.6 1.2 3 1.8 7.5-2.4 11.6z"
                          />
                        </svg>
                      </i>
                    </span>
                  </th>
                </tr>
              </thead>
              <tbody
                class="ant-table-tbody"
              >
                <tr
                  class="ant-table-row  ant-table-row-level-0"
                  data-row-key="0"
                >
                  <td
                    class="ant-table-column-has-filters"
                  >
                    <span
                      class="ant-table-row-indent indent-level-0"
                      style="padding-left:0px"
                    />
                    Jack
                  </td>
                </tr>
                <tr
                  class="ant-table-row  ant-table-row-level-0"
                  data-row-key="1"
                >
                  <td
                    class="ant-table-column-has-filters"
                  >
                    <span
                      class="ant-table-row-indent indent-level-0"
                      style="padding-left:0px"
                    />
                    Lucy
                  </td>
                </tr>
                <tr
                  class="ant-table-row  ant-table-row-level-0"
                  data-row-key="2"
                >
                  <td
                    class="ant-table-column-has-filters"
                  >
                    <span
                      class="ant-table-row-indent indent-level-0"
                      style="padding-left:0px"
                    />
                    Tom
                  </td>
                </tr>
                <tr
                  class="ant-table-row  ant-table-row-level-0"
                  data-row-key="3"
                >
                  <td
                    class="ant-table-column-has-filters"
                  >
                    <span
                      class="ant-table-row-indent indent-level-0"
                      style="padding-left:0px"
                    />
                    Jerry
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table.filter renders menu correctly 1`] = `
<div>
  <div
    class="ant-dropdown  ant-dropdown-placement-bottomLeft  ant-dropdown-hidden"
  >
    <div
      class="ant-table-filter-dropdown"
    >
      <ul
        class="ant-dropdown-menu  ant-dropdown-menu-root ant-dropdown-menu-vertical"
        role="menu"
        tabindex="0"
      >
        <li
          class="ant-dropdown-menu-item"
          role="menuitem"
        >
          <label
            class="ant-checkbox-wrapper"
          >
            <span
              class="ant-checkbox"
            >
              <input
                class="ant-checkbox-input"
                type="checkbox"
              />
              <span
                class="ant-checkbox-inner"
              />
            </span>
          </label>
          <span>
            Boy
          </span>
        </li>
        <li
          class="ant-dropdown-menu-item"
          role="menuitem"
        >
          <label
            class="ant-checkbox-wrapper"
          >
            <span
              class="ant-checkbox"
            >
              <input
                class="ant-checkbox-input"
                type="checkbox"
              />
              <span
                class="ant-checkbox-inner"
              />
            </span>
          </label>
          <span>
            Girl
          </span>
        </li>
        <li
          class="ant-dropdown-menu-submenu ant-dropdown-menu-submenu-vertical"
          role="menuitem"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-dropdown-menu-submenu-title"
            title="Title"
          >
            Title
            <i
              class="ant-dropdown-menu-submenu-arrow"
            />
          </div>
        </li>
      </ul>
      <div
        class="ant-table-filter-dropdown-btns"
      >
        <a
          class="ant-table-filter-dropdown-link confirm"
        >
          OK
        </a>
        <a
          class="ant-table-filter-dropdown-link clear"
        >
          Reset
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table.filter renders radio filter correctly 1`] = `
<div>
  <div
    class="ant-dropdown  ant-dropdown-placement-bottomLeft  ant-dropdown-hidden"
  >
    <div
      class="ant-table-filter-dropdown"
    >
      <ul
        class="ant-dropdown-menu  ant-dropdown-menu-root ant-dropdown-menu-vertical"
        role="menu"
        tabindex="0"
      >
        <li
          class="ant-dropdown-menu-item"
          role="menuitem"
        >
          <label
            class="ant-radio-wrapper"
          >
            <span
              class="ant-radio"
            >
              <input
                class="ant-radio-input"
                type="radio"
              />
              <span
                class="ant-radio-inner"
              />
            </span>
          </label>
          <span>
            Boy
          </span>
        </li>
        <li
          class="ant-dropdown-menu-item"
          role="menuitem"
        >
          <label
            class="ant-radio-wrapper"
          >
            <span
              class="ant-radio"
            >
              <input
                class="ant-radio-input"
                type="radio"
              />
              <span
                class="ant-radio-inner"
              />
            </span>
          </label>
          <span>
            Girl
          </span>
        </li>
        <li
          class="ant-dropdown-menu-submenu ant-dropdown-menu-submenu-vertical"
          role="menuitem"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-dropdown-menu-submenu-title"
            title="Title"
          >
            Title
            <i
              class="ant-dropdown-menu-submenu-arrow"
            />
          </div>
        </li>
      </ul>
      <div
        class="ant-table-filter-dropdown-btns"
      >
        <a
          class="ant-table-filter-dropdown-link confirm"
        >
          OK
        </a>
        <a
          class="ant-table-filter-dropdown-link clear"
        >
          Reset
        </a>
      </div>
    </div>
  </div>
</div>
`;
